compiler/rustc_metadata/src/locator.rs

```diff
@@ -723,18 +733,49 @@ impl<'a> CrateLocator<'a> {
             rejected_via_kind: self.rejected_via_kind,
             rejected_via_version: self.rejected_via_version,
             rejected_via_filename: self.rejected_via_filename,
+            rejected_via_invalid: self.rejected_via_invalid,
         })
     }
 }
 
+enum MetadataError<'a> {
+    /// The file was missing.
+    NotPresent(&'a Path),
+    /// The file was present and valid, but built by a different and incompatible version of the compiler.
+    WrongVersion(&'a Path),
+    /// The file was present and invalid.
+    LoadFailure(String),
+}
+
+impl fmt::Display for MetadataError<'_> {
```

Comment:
Could you move `MetadataError` and this impl to the bottom of the file where all other error reporting code lives?